### PR TITLE
Fix stream hanging upon corrupted gzipped data

### PR DESCRIPTION
**Pull request by ophirorly**
If the gzip header is corrupted, anyone trying to read from it will get stuck on the PipedInputStream.awaitSpace() function
This is a sample stack trace for such a case:

First we get an exception on the corrupted header:
`java.util.zip.ZipException: Not in GZIP format
	at java.util.zip.GZIPInputStream.readHeader(GZIPInputStream.java:165)
	at java.util.zip.GZIPInputStream.<init>(GZIPInputStream.java:79)
	at java.util.zip.GZIPInputStream.<init>(GZIPInputStream.java:91)
	at rawhttp.core.body.encoding.GZipUncompressorOutputStream.lambda$startReader$0(GZipUncompressorOutputStream.java:54)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)`

Then later, when trying to read the stream (via the LazyBodyReader) we get:
` java.lang.Thread.State: TIMED_WAITING (on object monitor)
        at java.lang.Object.wait(Native Method)
        at java.io.PipedInputStream.awaitSpace(PipedInputStream.java:273)
        at java.io.PipedInputStream.receive(PipedInputStream.java:231)
        - locked <0x00000006d2ba5eb0> (a java.io.PipedInputStream)
        at java.io.PipedOutputStream.write(PipedOutputStream.java:149)
        at rawhttp.core.body.encoding.GZipUncompressorOutputStream.write(GZipUncompressorOutputStream.java:47)
        at java.io.FilterOutputStream.write(FilterOutputStream.java:97)
        at rawhttp.core.body.BodyConsumer$ChunkedBodyConsumer.lambda$consumeDataInto$2(BodyConsumer.java:102)
        at rawhttp.core.body.BodyConsumer$ChunkedBodyConsumer$$Lambda$46/1788247731.accept(Unknown Source)
        at rawhttp.core.body.ChunkedBodyParser.parseChunkedBody(ChunkedBodyParser.java:104)
        at rawhttp.core.body.BodyConsumer$ChunkedBodyConsumer.consumeDataInto(BodyConsumer.java:101)
        at rawhttp.core.body.BodyReader.writeDecodedTo(BodyReader.java:115)
        at rawhttp.core.body.BodyReader.decodeBody(BodyReader.java:200)
        at rawhttp.core.body.BodyReader.decodeBodyToString(BodyReader.java:212)`

### CODE REVIEW

**Comment by renatoathaydes** [comment]
Hi @ophirorly . Looks like a great improvement,  but without a test, I can't know if this really fixes the problem. Can you write one?

I wrote all core tests in Kotlin, hope you're familiar with it? If not, I don't mind if you add a Java test.

You can look at existing tests like `rawhttp.core.body.FileBodyTest` and `rawhttp.core.body.InputStreamChunkDecoderTest`... I didn't add a test for gzip because I was just delegating that to the Java gzip stream, but you can create one new test class just for this.

**Comment by ophirorly** [comment]
Hi @renatoathaydes. Of course. I'll add a test when I'm able. Also - I'm thinking of saving the exception, and throwing it on finishDecoding() so that the caller will have an indication he's got a corrupted body. 
What do you think?

**Comment by renatoathaydes** [comment]
@ophirorly yeah, when the caller tries to get the body, it should definitely cause an Exception in this case.

**Comment by ophirorly** [comment]
@renatoathaydes I've added code to throw an exception if the parsed body was corruped, and two tests for both scenarios

File: rawhttp-core/src/test/kotlin/rawhttp/core/body/encoding/BodyDecodingTest.kt [hunk 1]
```diff
@@ -125,4 +126,28 @@ class BodyDecodingTest {
...
 
+    @Test
+    fun corruptedChunkedGzippedBodyDoesNotHang() {
+        val gzippedFileStream = BodyDecodingTest::class.java.getResourceAsStream("corrupted-chunked-and-gzipped-response.http")
+        val response = RawHttp().parseResponse(gzippedFileStream)
+        try {
+            response.body.get().decodeBodyToString(StandardCharsets.UTF_8)
+        } catch(e: IOException) {}
+
+        true shouldBe true
```
Comment (renatoathaydes) [inline rawhttp-core/src/test/kotlin/rawhttp/core/body/encoding/BodyDecodingTest.kt:137]: Thanks for the effort... It helps a lot that you've written a little test, but I have to say that the tests, as they are, are really not good enough... sorry to be blunt.

We need to check how long it took for the test to give up, what the Exception type was, what the message was...

I will accept your changes but will write more tests myself... I will push my code to this PR (hopefully, if I have permission) so we can discuss later if you would like.

Thanks again for finding the bug and contributing a first solution!

**Comment by renatoathaydes** [comment]
@ophirorly  I've pushed changes to the tests and then fixed the problems I found... I wanted to make sure the caller of `getDecodedBody` actually gets the appropriate Exception, which in this case is the Exception related to the GZIP Stream.
The second test you wrote is still failing, and I am not sure why... if you want to have a look, feel free to push a fix :) otherwise I will continue tomorrow.

File: rawhttp-core/src/main/java/rawhttp/core/body/BodyReader.java [hunk 2]
```diff
@@ -1,19 +1,16 @@
...
 
-import java.io.ByteArrayOutputStream;
-import java.io.Closeable;
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.OutputStream;
-import java.nio.charset.Charset;
-import java.util.Iterator;
-import java.util.Optional;
-import java.util.OptionalLong;
```
Comment (renatoathaydes) [inline rawhttp-core/src/main/java/rawhttp/core/body/BodyReader.java:11]: I will revert import changes later! My IDE lost my import settings apparently.... I avoid `*` imports as I've had many bugs caused by the wrong classes being imported accidentally.

File: rawhttp-core/src/test/kotlin/rawhttp/core/body/encoding/BodyDecodingTest.kt (hunk 1 above)
Comment (ophirorly) [reply rawhttp-core/src/test/kotlin/rawhttp/core/body/encoding/BodyDecodingTest.kt]: Thank you for the input. You have permission to the repository.
About the thrown exception - this PR fixes two different scenarios, each throws a different exception.
1. In the chunked gzipped case, the piped input stream is stuck, so when we close it - it throws an IOException with the message "Pipe closed".
2. In the non chunked corrupted gzip, it throws an IOException with the message "Not in GZIP format" (with my patch, it throws it on finishDecoding(), previously it was suppressed)

I thought about wrapping the call to write() with a try/catch and then wrapping the exception with a different one - but I thought it would make the code more cumbersome (need to check for that specific message, and then assume that it is that edge case). Considering that, I thought that it is better as it is - but I guess opinions may vary on this issue. 
The test could be modified to expect the "pipe closed" - but I think that's an internal implementation issue. The main thing is that an IOException was thrown, and that the flow is not stuck.

Regarding how long it took for the test to give up - I'm adding code to test it. 

I'll also change the non-chunked corrupted gzip test to verify that the exception message is "Not in GZIP format"

If you have any more ideas for tests - go ahead. Of course.

**Comment by ophirorly** [comment]
I think that the second test fails because you throw the WrappedException in the executor thread, which does nothing. 
You need to save it and then throw it (or a new exception) where the caller would get it. I used the finishDecoding(), since it guarantees no race conditions, and the user have to call it.

File: rawhttp-core/src/main/java/rawhttp/core/body/encoding/GZipUncompressorOutputStream.java [hunk 3]
```diff
@@ -69,23 +70,42 @@ public void flush() throws IOException {
...
+            closeQuietly(encodedBytesSink);
             readerExecution.get(5, TimeUnit.SECONDS);
         } catch (InterruptedException e) {
-            executorService.shutdownNow();
             throw new RuntimeException(e);
         } catch (ExecutionException e) {
-            executorService.shutdownNow();
-            throw new RuntimeException(e.getCause());
+            Throwable cause = e.getCause();
+            if (cause instanceof WrappedException) {
```
Comment (ophirorly) [inline rawhttp-core/src/main/java/rawhttp/core/body/encoding/GZipUncompressorOutputStream.java:81]: I think that currently, this flow will not be reached. The WrappedException is only thrown at the reader thread.

File: rawhttp-core/src/main/java/rawhttp/core/body/encoding/GZipUncompressorOutputStream.java [hunk 4]
```diff
@@ -69,23 +70,42 @@ public void flush() throws IOException {
...
+            }
+            throw new IOException(cause);
         } catch (TimeoutException e) {
-            executorService.shutdownNow();
             throw new RuntimeException("Timeout waiting for stream to close");
+        } finally {
+            closeQuietly(encodedBytesReceiver);
+            closeQuietly(encodedBytesSink);
+
+            executorService.shutdownNow();
```
Comment (ophirorly) [inline rawhttp-core/src/main/java/rawhttp/core/body/encoding/GZipUncompressorOutputStream.java:91]: I would simply save the exception from the reader thread and throw it here.

**Comment by ophirorly** [comment]
> @ophirorly I've pushed changes to the tests and then fixed the problems I found... I wanted to make sure the caller of `getDecodedBody` actually gets the appropriate Exception, which in this case is the Exception related to the GZIP Stream.
> The second test you wrote is still failing, and I am not sure why... if you want to have a look, feel free to push a fix :) otherwise I will continue tomorrow.

I think it fails because the caller thread is not stuck on writing. It already wrote all the data to the stream, and then called finishDecoding(). When you interrupt the thread - it catches the interrupted exception, instead of the execution result.

**Comment by ophirorly** [comment]
@renatoathaydes I've add a fix for the second issue. 
Personally, I would use the caughtException mechanism for all cases, instead of the WrappedException, but I guess opinions will vary on this issue

**Comment by renatoathaydes** [comment]
> The problem is that if the user thread is stuck on writing we'd want to interrupt it, but if it is waiting in finishDecoding() 

@ophirorly right, I'd forgotten to close the stream, which would've called `finishDecoding`... thanks for finding that.

> so I simply updated an error state ('caughtException')

The problem with that is that when you have multiple threads, updating a field like that is unsafe and prone to race conditions... But you're right: keeping a reference to the error seems to be the only way to re-throw it later... I've improved your fix a bit to use `AtomicReference`, and I think it's safe from race conditions now because the only way for us to get an `InterruptedException` at the `finishDecoding` method is if the reader thread interrupted the writer Thread that was waiting... and it only does that immediately after setting the reference to the error (I am surprised your fix worked because you updated a non-volatile field and immediately read it from another Thread, but the JVM seems very lenient regarding multi-threaded access to fields like this - just don't count on it :) ).

We actually needed to both keep a ref to the Error AND do as I did, interrupt the reader (the way you were doing before worked but the caller would get a misleading `Pipe closed` error) to stop it from hanging... I wish it were simpler, but this now seems to cover everything.

All tests are passing so I will merge as soon as I get a CI build green light.

Thanks for the contribution.

**Comment by renatoathaydes** [comment]
Looks like the CI is not building because this is a fork! Will merge and fix anything if necessary.

**Comment by ophirorly** [comment]
>  (I am surprised your fix worked because you updated a non-volatile field and immediately read it from another Thread, but the JVM seems very lenient regarding multi-threaded access to fields like this - just don't count on it :) ).

I think that in this scenario it's ok, because I would expect the JVM to put a memory barrier at that point (since the thread made a blocking call). However, putting a volatile is definitely better practice, and so is AtomicRef.

I also don't think there's a risk of significant race conditions - there is one reader thread that can only update the exception once, finishDecoding should also be called once, and reading the exception happens once, and it's after the reader has finished one way or another. 
But again - I agree that even if I'm right - putting an AtomicReference is the better practice here.
 
> We actually needed to both keep a ref to the Error AND do as I did, interrupt the reader (the way you were doing before worked but the caller would get a misleading `Pipe closed` error) to stop it from hanging... I wish it were simpler, but this now seems to cover everything.

Yes it does. 
Thank you for perfecting it and taking the time to explain it so throughly.
